lua/null-ls/builtins/diagnostics.lua

```diff
@@ -201,128 +201,157 @@ M.chktex = h.make_builtin({
     generator_opts = {
-        command = "write-good",
-        args = { "--text=$TEXT", "--parse" },
+        command = "clang-tidy",
+        args = { "--quiet", "$FILENAME" },
+        to_stdin = false,
+        to_temp_file = true,
+        format = "line",
+        on_output = from_pattern(
+            "(%d+):(%d+): (%w+): (.*) %[(.*)%]", --
+            { "row", "col", "severity", "message", "code" }
+        ),
+    },
+    factory = h.generator_factory,
+})
+
+M.eslint = h.make_builtin({
+    method = DIAGNOSTICS,
+    filetypes = { "javascript", "javascriptreact", "typescript", "typescriptreact", "vue" },
+    generator_opts = {
+        command = "eslint",
+        args = { "-f", "json", "--stdin", "--stdin-filename", "$FILENAME" },
+        to_stdin = true,
+        format = "json_raw",
+        check_exit_code = function(code)
+            return code <= 1
+        end,
+        use_cache = true,
+        on_output = function(params)
+            params.messages = params.output and params.output[1] and params.output[1].messages or {}
+            if params.err then
+                table.insert(params.messages, { message = params.err })
+            end
+
+            local parser = from_json({
+                attributes = {
+                    severity = "severity",
+                },
+                severities = {
+                    default_severities["warning"],
+                    default_severities["error"],
+                },
+            })
+
+            return parser({ output = params.messages })
+        end,
+    },
+    factory = h.generator_factory,
+})
+
+M.flake8 = h.make_builtin({
+    method = DIAGNOSTICS,
+    filetypes = { "python" },
+    generator_opts = {
+        command = "flake8",
+        to_stdin = true,
+        to_stderr = true,
+        args = { "--stdin-display-name", "$FILENAME", "-" },
         format = "line",
         check_exit_code = function(code)
             return code == 0 or code == 255
         end,
         on_output = from_pattern(
-            [[(%d+):(%d+):("([%w%s]+)".*)]], --
-            { "row", "col", "message", "_quote" },
+            [[:(%d+):(%d+): (([EFW])%w+) (.*)]], --
+            { "row", "col", "code", "severity", "message" },
             {
-                adapters = { diagnostic_adapters.end_col.from_quote },
-                offsets = { col = 1, end_col = 1 },
+                severities = {
+                    E = default_severities["error"],
+                    W = default_severities["warning"],
+                    F = default_severities["information"],
+                },
             }
         ),
     },
     factory = h.generator_factory,
 })
 
-M.markdownlint = h.make_builtin({
+M.golangci_lint = h.make_builtin({
     method = DIAGNOSTICS,
-    filetypes = { "markdown" },
+    filetypes = { "go" },
     generator_opts = {
-        command = "markdownlint",
-        args = { "--stdin" },
+        command = "golangci-lint",
+        args = { "run" },
```

Comment:
```suggestion
        command = "golangci-lint",
        args = { "run", "$FILENAME"},
```
Without `$FILENAME` golangci-lint reports errors for the entire projec. But they will show up in the diagnostics for the current file, creating a lot of false positives.

Reply:
Running `golangci-lint run $FILENAME` will produce diagnostics for only that file, but unfortunately without intelligence from other files.

As an easy to reproduce example:

```sh
mkdir example && cd example
go mod init example
echo -e "package main\n\nfunc main() {}" > main_1.go
echo -e "package main\n\nfunc main() {}" > main_2.go
```

Running `golangci-lint run` produces:
```
main_2.go:3:6: `main` redeclared in this block (typecheck)
func main() {}
     ^
main_1.go:3:6: other declaration of main (typecheck)
func main() {}
     ^
```

But running `golangci-lint run main_1.go` does not find the error, because it is not aware of the other files in the project.

The ideal solution would be being able to run it at the package level, then filter down to only report errors from the open file.

Reply:
Filtering by filename should be simple. [The original feature request](https://github.com/jose-elias-alvarez/null-ls.nvim/issues/62) mentions speed concerns, so I wonder about performance. null-ls updates diagnostics according to LSP behavior, but LSP behavior assumes we're sending updates for one file at a time to a language server, so slow-running, project-wide linters could cause issues.

Reply:
There are some things I don't understand, and that maybe you could explain.
if I run `cat cmd/main.go | golangci-lint run` it does work, doesn't it mean that it supports stdin ?